### ISSUE TITLE
Can no longer slice csr_matrix in 1.3.0

### ISSUE THREAD

**Issue by dfrankow** [scipy/scipy#11200]
I have a code example that works in scipy 1.1.0, 1.2.0, 1.2.1 and has a stack trace in 1.3.0, 1.3.1, 1.3.2.

#### Reproducing code example:

```
import pandas as pd
from scipy.sparse import csr_matrix

df = pd.DataFrame({'thing': [1, 1, 2, 2, 2],
                   'score': [0.12, 0.13, 0.14, 0.15, 0.17]})

row_index = df['thing'] == 1
print(type(row_index), row_index)
print(df[row_index])
sdf = csr_matrix(df)
print(sdf[row_index])
```

#### Error message:

```
Traceback (most recent call last):
  File "./foo.py", line 13, in <module>
    print(sdf[row_index])
  File "root/.venv/lib/python3.7/site-packages/scipy/sparse/_index.py", line 59, in __getitem__
    return self._get_arrayXslice(row, col)
  File "root/.venv/lib/python3.7/site-packages/scipy/sparse/csr.py", line 325, in _get_arrayXslice
    return self._major_index_fancy(row)._get_submatrix(minor=col)
  File "root/.venv/lib/python3.7/site-packages/scipy/sparse/compressed.py", line 690, in _major_index_fancy
    np.cumsum(row_nnz[idx], out=res_indptr[1:])
  File "<__array_function__ internals>", line 6, in cumsum
  File "root/.venv/lib/python3.7/site-packages/numpy/core/fromnumeric.py", line 2423, in cumsum
    return _wrapfunc(a, 'cumsum', axis=axis, dtype=dtype, out=out)
  File "root/.venv/lib/python3.7/site-packages/numpy/core/fromnumeric.py", line 61, in _wrapfunc
    return bound(*args, **kwds)
ValueError: provided out is the wrong size for the reduction
```

#### Output of working version (on earlier versions of scipy):

```
<class 'pandas.core.series.Series'> 0     True
1     True
2    False
3    False
4    False
Name: thing, dtype: bool
   thing  score
0      1   0.12
1      1   0.13
  (0, 0)	1.0
  (0, 1)	0.13
  (1, 0)	1.0
  (1, 1)	0.13
  (2, 0)	1.0
  (2, 1)	0.12
  (3, 0)	1.0
  (3, 1)	0.12
  (4, 0)	1.0
  (4, 1)	0.12
```

#### Scipy/Numpy/Python version information:
```
1.3.0 1.17.3 sys.version_info(major=3, minor=7, micro=4, releaselevel='final', serial=0)
```

See the same info on stackoverflow [here](https://stackoverflow.com/questions/59273855/how-to-select-only-some-rows-of-a-scipy-sparse-csr-matrix).

I'm running in OS X, tried pandas 0.25 and 0.25.3, though I don't know that those matter.

**Comment by dfrankow** [comment]
[This issue](https://github.com/scipy/scipy/issues/9639) may or may not be related.

**Comment by perimosocordiae** [comment]
I can confirm this issue is present on master (1.5.0.dev0+0a7b39a).

**Comment by perimosocordiae** [comment]
The issue here stems from the use of a pandas Series object as a boolean/logical index array. We check for boolean indices here:

https://github.com/scipy/scipy/blob/0a7b39a3973bf0ed95a497764445ccbaed7e990b/scipy/sparse/_index.py#L279

Since `pandas.Series` is not a subclass of `numpy.ndarray`, the indices are not converted to integers, breaking the assumptions in the code that follows.

You can see this in the posted example:

```
In [29]: sdf[row_index.values]                                                  
Out[29]: 
<2x2 sparse matrix of type '<class 'numpy.float64'>'
	with 4 stored elements in Compressed Sparse Row format>
```

To see that the change is taking place in index validation:

```
In [30]: sdf._validate_indices((row_index, slice(None)))                        
Out[30]: (array([ True,  True, False, False, False]), slice(None, None, None))

In [31]: sdf._validate_indices((row_index.values, slice(None)))                 
Out[31]: (array([0, 1]), slice(None, None, None))
```

So for the time being, you can work around the issue by changing index arrays from Series to plain numpy arrays. A fix for the index validation to be less strict about types when checking for logical index arrays would solve this for good, though.

**Comment by dfrankow** [comment]
Thanks!

**Comment by ananyashreyjain** [comment]
Instead of checking index array for `numpy.ndarray` wouldn't it be better if check for the presence of attribute `ndim`?  Pandas Series object already has attribute `ndim`. We may get the desired leniency in array type by making this change.

**Comment by ewmoore** [comment]
Looks like this was fixed in #11265.  Closing.
